core/src/core/pmmr/segment.rs

```diff
@@ -0,0 +1,519 @@
+	}
+
+	/// Offset (in leaf idx) of first leaf in the segment
+	#[inline]
+	fn leaf_offset(&self) -> u64 {
+		self.identifier.idx * self.segment_capacity()
+	}
+
+	/// Number of leaves in this segment. Equal to capacity except for the final segment, which can be smaller
+	#[inline]
+	fn segment_unpruned_size(&self, last_pos: u64) -> u64 {
+		min(
+			self.segment_capacity(),
+			pmmr::n_leaves(last_pos).saturating_sub(self.leaf_offset()),
+		)
+	}
+
+	/// Whether the segment is full (size == capacity)
+	#[inline]
+	fn full_segment(&self, last_pos: u64) -> bool {
+		self.segment_unpruned_size(last_pos) == self.segment_capacity()
+	}
+
+	/// Inclusive range of MMR positions for this segment
+	#[inline]
+	fn segment_pos_range(&self, last_pos: u64) -> (u64, u64) {
+		let segment_size = self.segment_unpruned_size(last_pos);
+		let leaf_offset = self.leaf_offset();
+		let first = pmmr::insertion_to_pmmr_index(leaf_offset + 1);
+		let last = if self.full_segment(last_pos) {
+			pmmr::insertion_to_pmmr_index(leaf_offset + segment_size)
+				+ (self.identifier.log_size as u64)
+		} else {
+			last_pos
+		};
+
+		(first, last)
+	}
+
+	fn get_hash(&self, pos: u64) -> Result<Hash, SegmentError> {
+		self.hashes
+			.binary_search_by(|&(p, _)| p.cmp(&pos))
+			.map(|p| self.hashes[p].1)
+			.map_err(|_| SegmentError::MissingHash(pos))
+	}
+}
+
+impl<T> Segment<T>
+where
+	T: Readable + Writeable + Debug,
+{
+	/// Generate a segment from a PMMR
+	pub fn from_pmmr<U, B>(
+		segment_id: SegmentIdentifier,
+		pmmr: &ReadonlyPMMR<'_, U, B>,
+		prunable: bool,
+	) -> Result<Self, SegmentError>
+	where
+		U: PMMRable<E = T>,
+		B: Backend<U>,
+	{
+		let mut segment = Segment {
+			identifier: segment_id,
+			hashes: Vec::new(),
+			leaf_data: Vec::new(),
+			proof: SegmentProof::empty(),
+		};
+
+		let last_pos = pmmr.unpruned_size();
+		if segment.segment_unpruned_size(last_pos) == 0 {
+			return Err(SegmentError::NonExistent);
+		}
+
+		// Fill leaf data and hashes
+		let (segment_first_pos, segment_last_pos) = segment.segment_pos_range(last_pos);
+		for pos in segment_first_pos..=segment_last_pos {
+			if pmmr::is_leaf(pos) {
+				if let Some(data) = pmmr.get_data_from_file(pos) {
+					segment.leaf_data.push((pos, data));
+					continue;
+				} else if !prunable {
+					return Err(SegmentError::MissingLeaf(pos));
+				}
+			}
+			// TODO: optimize, no need to send every intermediary hash
+			if prunable {
+				if let Some(hash) = pmmr.get_from_file(pos) {
+					segment.hashes.push((pos, hash));
+				}
+			}
+		}
+
+		if segment.leaf_data.is_empty() {
+			return Err(SegmentError::Pruned);
+		}
+
+		// Segment merkle proof
+		segment.proof =
+			SegmentProof::generate(pmmr, last_pos, segment_first_pos, segment_last_pos)?;
+
+		Ok(segment)
+	}
+}
+
+impl<T> Segment<T>
+where
+	T: PMMRIndexHashable,
+{
+	/// Calculate root hash of this segment
+	pub fn root(&self, last_pos: u64, bitmap: Option<&Bitmap>) -> Result<Hash, SegmentError> {
```

Comment:
This root calculation is good for now but it feels like longer term we want/need a more flexible (sub)PMMR interface and actually (re)build a (sub)PMMR up given the provided hashes and leaf data.

The segment is a subtree (potentially multiple smaller subtrees) in a larger PMMR. And the root is the root of this subtree.

If we could start at some leaf offset (rather than pos 1) and simply build a partial/sub PMMR from that point on then we effectively get the root for free.

We do assume we only ever add individual leaves to a PMMR currently. We would need to be able to add parent hashes directly, skipping pruned leaves, in some more flexible way to accomplish this.

I'm not suggesting we do this for this PR, just something for us to think about in the future.